x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -710,49 +711,171 @@ private static void writeMetadata(StreamOutput out, Map<String, Object> metadata
         }
+        checkNoRunAs(this, "Internal");
+    }
 
-        assert effectiveSubject != null;
-        assert authenticatingSubject != null;
-        assert type != null;
-        assert effectiveSubject.getTransportVersion().equals(authenticatingSubject.getTransportVersion());
+    private void checkConsistencyForApiKeyAuthenticationType() {
+        final RealmRef authenticatingRealm = authenticatingSubject.getRealm();
+        if (false == authenticatingRealm.isApiKeyRealm() && false == authenticatingRealm.isRemoteAccessRealm()) {
+            throw new IllegalArgumentException(
+                Strings.format("API key authentication cannot have realm type [%s]", authenticatingRealm.type)
+            );
+        }
+        checkNoDomain(authenticatingRealm, "API key");
+        checkNoInternalUser(authenticatingSubject, "API key");
+        checkNoRole(authenticatingSubject, "API key");
+        if (authenticatingSubject.getMetadata().get(AuthenticationField.API_KEY_ID_KEY) == null) {
+            throw new IllegalArgumentException("API key authentication requires metadata to contain a non-null API key ID");
+        }
+        if (Subject.Type.REMOTE_ACCESS == authenticatingSubject.getType()) {
+            if (authenticatingSubject.getMetadata().get(REMOTE_ACCESS_AUTHENTICATION_KEY) == null) {
+                throw new IllegalArgumentException(
+                    "Remote access authentication requires metadata to contain a non-null serialized remote access authentication"
+                );
+            }
+            if (authenticatingSubject.getMetadata().get(REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY) == null) {
+                throw new IllegalArgumentException(
+                    "Remote access authentication requires metadata to contain a non-null serialized remote access role descriptors"
+                );
+            }
+            checkNoRunAs(this, "Remote access");
+        } else {
+            if (isRunAs()) {
+                checkRunAsConsistency(effectiveSubject, authenticatingSubject);
+            }
+        }
+    }
 
+    private void checkConsistencyForRealmAuthenticationType() {
+        if (Subject.Type.USER != authenticatingSubject.getType()) {
+            throw new IllegalArgumentException("Realm authentication must have subject type of user");
+        }
         if (isRunAs()) {
-            assert authenticatingSubject != effectiveSubject : "isRunAs logic does not hold";
-            assert false == User.isInternal(effectiveSubject.getUser()) && false == User.isInternal(authenticatingSubject.getUser())
-                : "internal users cannot participate in run-as";
+            checkRunAsConsistency(effectiveSubject, authenticatingSubject);
+        }
+    }
+
+    private void checkConsistencyForTokenAuthenticationType() {
+        final RealmRef authenticatingRealm = authenticatingSubject.getRealm();
+        assert false == authenticatingRealm.isAttachRealm()
```

Comment:
I'm wondering if this should be an exception instead of an assertion.

Reply:
This does not hold for custom realms since they can have type and name identical to synthetic realms. Custom realms with these names are likely to fail in other places. But we haven't failed them in name/type checks yet (see also `hasSyntheticRealmNameOrType`). So I decided to leave it as an assertion. I'll add a comment above it for explanation.